packages/core/src/api/Editor.ts

```diff
@@ -0,0 +1,416 @@
+    const content: Node[] = [];
+
+    for (const styledText of block.styledTextContent) {
+      const marks = [];
+
+      for (const style of styledText.styles) {
+        marks.push(this.tiptapEditor.schema.mark(style.type, style.props));
+      }
+
+      content.push(this.tiptapEditor.schema.text(styledText.text, marks));
+    }
+
+    const contentNode = this.tiptapEditor.state.schema.nodes[block.type].create(
+      block.props,
+      content
+    );
+
+    const children: Node[] = [];
+
+    if (block.children) {
+      for (const child of block.children) {
+        children.push(this.blockToNode(child));
+      }
+    }
+
+    const groupNode = this.tiptapEditor.state.schema.nodes["blockGroup"].create(
+      {},
+      children
+    );
+
+    return this.tiptapEditor.state.schema.nodes["blockContainer"].create(
+      {},
+      children.length > 0 ? [contentNode, groupNode] : contentNode
+    );
+  }
+
+  private nodeToBlock(node: Node): Block {
+    const cachedBlock = this.blockCache.get(node);
+    if (cachedBlock) {
+      return cachedBlock;
+    }
+
+    const blockInfo = getBlockInfoFromPos(node, 0)!;
+
+    const styledText: StyledText[] = [];
+    blockInfo.contentNode.content.forEach((node) => {
+      const styles: Style[] = [];
+
+      for (const mark of node.marks) {
+        styles.push({
+          type: mark.type.name,
+          props: mark.attrs,
+        } as Style);
+      }
+
+      styledText.push({
+        text: node.textContent,
+        styles: styles,
+      });
+    });
+
+    const children: Block[] = [];
+    for (let i = 0; i < blockInfo.numChildBlocks; i++) {
+      children.push(this.nodeToBlock(blockInfo.node.lastChild!.child(i)));
+    }
+
+    const block = {
+      id: blockInfo.id,
+      type: blockInfo.contentType.name,
+      // TODO: Only return attributes specified in the BlockSpec, not all attributes (e.g. ordered list item index should
+      //  not be included).
+      props: blockInfo.contentNode.attrs,
+      textContent: blockInfo.contentNode.textContent,
+      styledTextContent: styledText,
+      children: children,
+    } as Block;
+
+    this.blockCache.set(node, block);
+
+    return block;
+  }
+
+  public get allBlocks(): Block[] {
+    const blocks: Block[] = [];
+
+    this.tiptapEditor.state.doc.firstChild!.descendants((node) => {
+      blocks.push(this.nodeToBlock(node));
+
+      return false;
+    });
+
+    return blocks;
+  }
+
+  public get cursorPosition(): CursorPosition {
+    const { node } = getBlockInfoFromPos(
+      this.tiptapEditor.state.doc,
+      this.tiptapEditor.state.selection.from
+    )!;
+
+    return { block: this.nodeToBlock(node) };
+  }
+
+  public insertBlocks(
+    blocksToInsert: Block[],
+    referenceBlockId: string,
+    placement?: "before" | "after" | "nested"
+  ): void {
+    if (!placement) {
+      placement = "before";
+    }
+
+    let blockFound = false;
+
+    this.tiptapEditor.state.doc.descendants((node, pos) => {
```

Comment:
don't we have (or can we make?) a separate function that just get's the tiptap node by ID?